### PR TITLE
WT-9030 Add a new undocumented configuration string remove_shared to WT_SESSION::drop

### PR DESCRIPTION
**Pull request by mm-ng**
The current implementation only supports `dir_store` and does not support `S3`.

Also, I'm not sure if there is a better way to generate all the shared object files - I'm currently generating them manually as the fs_directory_list function does not work as I thought it would on the `tiered->bstorage->bucket directory`.

### CODE REVIEW

**Comment by sueloverso** [comment]
@mm-ng I took a look at this PR and have some suggestions for the direction it should take. I am going to add them to the ticket so they're preserved in that.

**Comment by sueloverso** [comment]
@mm-ng This is looking closer to me. I think the right place to call the new `wt_meta_track_drop_object` is the same loop in `drop_tiered` where we call `wt_meta_track_drop` if the configuration is set for removing the shared object. 

Also realize that even when remove_shared is set, the object might not yet exist in the bucket. So the extension changes for removing from the cache and/or bucket should ignore errors that the object does not exist such as ENOENT in the dir_store.

**Comment by mm-ng** [comment]
> @mm-ng This is looking closer to me. I think the right place to call the new `wt_meta_track_drop_object` is the same loop in `drop_tiered` where we call `wt_meta_track_drop` if the configuration is set for removing the shared object.
> 
> Also realize that even when remove_shared is set, the object might not yet exist in the bucket. So the extension changes for removing from the cache and/or bucket should ignore errors that the object does not exist such as ENOENT in the dir_store.

For scenarios where the object does not exist, I'm not sure how we should indicate that the ENOENT return code is fine as `dir_store_remove` will call `__posix_fs_remove` if that is the correct function for the underlying file system. I think `__posix_fs_remove` expects ret to be 0 and I'm not sure if it is safe to change this function?

**Review by ddanderson — COMMENTED** [review]
I had a couple questions and a request for refactoring, otherwise it looked pretty good.

File: ext/storage_sources/dir_store/dir_store.c [hunk 1]
```diff
@@ -1090,15 +1090,62 @@ dir_store_rename(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *f
...
+        return (dir_store_err(FS2DS(file_system), session, ENOMEM, "dir_store_remove"));
+    if (snprintf(cache_filename, cache_file_len, "%s/%s", dir_store_fs->cache_dir, name) >=
+      (int)cache_file_len)
+        return (dir_store_err(FS2DS(file_system), session, EINVAL, "overflow sprintf"));
+
+    dir_store_exist(file_system, session, name, &existp);
+    if (existp) {
+        ret = wt_fs->fs_remove(wt_fs, session, bucket_filename, flags);
+        if (ret != 0 && ret != ENOENT)
+            goto err;
```
Comment (ddanderson) [inline ext/storage_sources/dir_store/dir_store.c:1138]: Might be good to call `dir_store_err` before the jump to `err` here, and pass it `bucket_filename`.  That way there will be an error message telling which file couldn't be removed.  Same thing after next remove call.

File: ext/storage_sources/dir_store/dir_store.c [hunk 2]
```diff
@@ -1090,15 +1090,62 @@ dir_store_rename(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *f
...
+    }
+
+    /* Generate the paths to the shared object files in the bucket and cache directories. */
+    bucket_file_len = strlen(dir_store_fs->bucket_dir) + strlen(name) + 2;
+    if ((bucket_filename = calloc(bucket_file_len, sizeof(char))) == NULL)
+        return (dir_store_err(FS2DS(file_system), session, ENOMEM, "dir_store_remove"));
+    if (snprintf(bucket_filename, bucket_file_len, "%s/%s", dir_store_fs->bucket_dir, name) >=
+      (int)bucket_file_len)
+        return (dir_store_err(FS2DS(file_system), session, EINVAL, "overflow sprintf"));
+
```
[the remark this answers is not in the document]
Comment (sueloverso) [reply ext/storage_sources/dir_store/dir_store.c:1126]: @ddanderson I put some suggestions into the ticket (for posterity and review in the future) that are similar to this. I had a couple other suggestions for the remove path that you might want to consider.

**Review by sueloverso — COMMENTED** [review]
I added some general comments in the ticket.

File: src/schema/schema_drop.c [hunk 3]
```diff
@@ -182,10 +182,15 @@ __drop_tiered(WT_SESSION_IMPL *session, const char *uri, bool force, const char
...
-    bool exist, remove_files;
+    bool exist, remove_files, remove_shared;
 
     WT_RET(__wt_config_gets(session, cfg, "remove_files", &cval));
     remove_files = cval.val != 0;
+    WT_RET(__wt_config_gets(session, cfg, "remove_shared", &cval));
+    remove_shared = cval.val != 0;
+
+    if (!remove_files && remove_shared)
+        WT_RET(EINVAL);
```
Comment (AlexeyAnisimovAU) [inline src/schema/schema_drop.c:193]: We probably should provide an error message here with `WT_RET_MSG`.

**Review by AlexeyAnisimovAU — COMMENTED** [review]
Great job! One comment from me.

File: ext/storage_sources/dir_store/dir_store.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mm-ng) [reply ext/storage_sources/dir_store/dir_store.c:1126]: Yes, I can definitely use that function here, will fix this in the next commit.

File: src/schema/schema_drop.c (hunk 3 above)
Comment (mm-ng) [reply src/schema/schema_drop.c:193]: Good suggestion, thanks!

File: test/suite/test_tiered16.py [hunk 4]
```diff
@@ -71,26 +92,14 @@ def test_remove_shared(self):
...
-                self.assertFalse(os.path.isfile("bucket1/pfx_tiereda-0000000002.wtobj"))
-                self.assertTrue(os.path.isfile("bucket1/pfx_tieredb-0000000001.wtobj"))
-                self.assertTrue(os.path.isfile("bucket1/pfx_tieredb-0000000002.wtobj"))
-
-                self.assertFalse(os.path.isfile("cache-bucket1/pfx_tiereda-0000000001.wtobj"))
-                self.assertFalse(os.path.isfile("cache-bucket1/pfx_tiereda-0000000002.wtobj"))
-                self.assertTrue(os.path.isfile("cache-bucket1/pfx_tieredb-0000000001.wtobj"))
-                self.assertTrue(os.path.isfile("cache-bucket1/pfx_tieredb-0000000002.wtobj"))
+            self.check_cache(cache_dir, [self.bucket_prefix + obj1file_b, self.bucket_prefix + obj2file_b])
+            self.check_bucket([self.bucket_prefix + obj1file_b, self.bucket_prefix + obj2file_b])
```
Comment (ddanderson) [inline test/suite/test_tiered16.py:96]: This is so much better!

File: ext/storage_sources/dir_store/dir_store.c [hunk 5]
```diff
@@ -1096,55 +1096,57 @@ static int
...
+            ret = dir_store_err(
+              dir_store_fs->dir_store, session, errno, "%s: dir_store_remove stat", bucket_path);
+            goto err;
+        }
+    } else {
+        if ((ret = wt_fs->fs_remove(wt_fs, session, bucket_path, flags)) != 0) {
+            ret = dir_store_err(
+              dir_store_fs->dir_store, session, errno, "%s: dir_store_remove", bucket_path);
             goto err;
+        }
```
Comment (ddanderson) [inline ext/storage_sources/dir_store/dir_store.c:1144]: This is much better.  I'd still like to see this refactored - nearly identical code to check the bucket vs check the cache directory.  Make a private function to do the repeated work.
Sorry to keep beating on refactoring common bits, but I think always think it's an important element of code quality.

**Review by ddanderson — APPROVED** [review]
I'm approving, with a suggestion for a little more refactoring.

File: ext/storage_sources/dir_store/dir_store.c [hunk 6]
```diff
@@ -1090,15 +1090,64 @@ dir_store_rename(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *f
 
 /*
  * dir_store_remove --
- *     POSIX remove, not supported for cloud objects.
+ *     Calls the corresponding remove method of the file system behind our dir_store storage source
```
Comment (sueloverso) [inline ext/storage_sources/dir_store/dir_store.c:1093]: Minor: End the sentence with a period.

**Review by sueloverso — APPROVED** [review]
LGTM. I have a few very minor style comments. This looks great now!

**Comment by mm-ng** [comment]
@ddanderson No worries, I agree with you on the refactoring, I've pulled out the logic into another function.

**Review by AlexeyAnisimovAU — APPROVED** [review]
Well done! LGTM

**Pull request merged**